### ISSUE TITLE
Add Substeps

### ISSUE THREAD

**Issue by gossi** [impress/impress.js#81]
Hey there,

I would like to see "substeps". I don't know how to call those properly, I give a little description:

Imagine you have your presentation with some slides/steps. Now on one slide you have a list and you want to fly in the list items one-by-one each by a arrow right keystroke. I tried it that way:

```
<ul>
<li class="step">Item 1</li>
<li class="step">Item 2</li>
</ul>
```

It didn't worked, of course I didn't expected that. But would be cool to have something like that.

Thanks very much.

**Comment by bioinfornatics** [comment]
@tehfoo  why you don not do a pull request ?

**Comment by patricebeutler** [comment]
I am working on a solution for this problem. Basically with this solution there is the possibility to add javascript functions as "steps". In this functions you can do whatever you want. Also, there are functions for reverse the changes done by these step-functions. When the work and the testing is done, i will file a pull-request or create a fork.

**Comment by Announcement** [comment]
Tween not steps pls

**Comment by patricebeutler** [comment]
@Announcement Thanks for the hint. My fault. I corrected it.

**Comment by patricebeutler** [comment]
Hey people. If you need this now take a look at my fork: https://github.com/patricebeutler/impress.js. Its exactly the same beside the supsteps-feature.

**Comment by calebeby** [comment]
Any chance this will be merged?

**Comment by naraesk** [comment]
Seems like @bartaz  doesn't have much time for this project recently. However, it would be great if a solution could be found that this repository stays at least maintained. There are several forks of it currently, can't use them simultaneously.

**Comment by henrikingo** [comment]
Hi @calebeby, @naraesk and others

Thanks for pinging on this thread again. Like bartaz, I kinda felt this is a silly powerpoint feature, so it wasn't top of my priority list last year. At the same time, it is true this must be the most requested feature at this point. So I have implemented it [in my fork](https://github.com/henrikingo/impress.js) today. ([Demo](http://henrikingo.github.io/impress.js/examples/classic-slides/#/step-9)).

> it would be great if a solution could be found that this repository stays at least maintained.
> There are several forks of it currently, can't use them simultaneously.

Just to let everyone know, my fork:

* Is actively maintained (for over a year now)
* Intended to be general purpose, so I add feature requests like this, even if I don't personally use them
* I'm happy to take pull requests too (please [familiarize yourself with the plugin framework](https://github.com/henrikingo/impress.js/blob/myfork/src/plugins/README.md))
* I originally added a plugin framework, based on helpful guidance from bartaz himself. Even today, he has at least not objected to what I ended up doing. In this sense, I've done my best to continue delivering upon the original vision of impress.js
* I periodically merge from upstream and have avoided doing some things in order to remain compatible with it. I realize upstream impress.js is stagnant, but should it ever spring to life again, it's been my hope that the work in my fork could be merged back upstream.

**Comment by calebeby** [comment]
@henrikingo Thanks for the update. That's good to hear that you are maintaining a fork. Thanks!

**Comment by ecolui** [comment]
@henrikingo - Thank you for the update.  I also did some work on the sub-steps feature that was commonly requested.  I did this by introducing two new functions (e.g. registerEnter and registerExit) that will allow the user to register any function as a sub-step inside a slide.  The ability to introduce user defined functions as sub-steps gives the user the flexibility to do whatever they want as a sub-step (e.g. growing/shrinking text, transition opacity from 0 to 1, spinning text, etc).  I committed the code to a fork of this project just for the sake of sharing this idea and the corresponding committed code (url provided below).  I am not going to actively maintain the forked version - i just forked the project in order to provide the following url.  

https://github.com/ecolui/impress.js/commit/004e8b6b1ba53dac4af31e7fb3cbcbf96cbd47ea

* Note: In the preceding url, pay attention to the new commented code at the bottom of index.html. That commented code shows you how to use the new registerEnter and registerExit functions.
